{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genetics Mapping Lab üåùüåªüåùüåªüåù\n",
    "\n",
    "*Created by Julia Gorman 4-18-19*\n",
    "\n",
    "**Purpose:** To generate data sets and recombinates genes and chromosome locations\n",
    "We only allow alleles ABCD in that precise order. \n",
    "\n",
    "Looking to create tool that randomizes the map distances if the user so chooses so that they can figure out the data themselves\n",
    " \n",
    " \n",
    "This will then send the professor an email of what the randomized data was so that the professor can see what positions the student got but the student cannot\n",
    "\n",
    "\n",
    "Interference: only affects double crossovers, multiple prob of double crossover by (1-n)\n",
    "\n",
    "## üß¨ üß¨ üß¨ üß¨ üß¨ üß¨ üß¨ üß¨ üß¨ üß¨ üß¨ üß¨ üß¨ üß¨ üß¨ üß¨ üß¨ üß¨ üß¨ üß¨  üß¨ üß¨ üß¨ üß¨ üß¨ üß¨ üß¨ üß¨ üß¨ üß¨ üß¨ üß¨"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "import random\n",
    "from tabulate import *\n",
    "\n",
    "def initialize():\n",
    "    print('Welcome to the genetic mapping simulator.')\n",
    "    print('What type of genetic cross would you like to observe?')\n",
    "    print('Option 1: 3-1 cross')\n",
    "    print('Press 1 for Option 1')\n",
    "    print('or')\n",
    "    print('Option 2: 2-2 cross')\n",
    "    print('Press 2 for Option 2')\n",
    "    crossType = int(input())\n",
    "\n",
    "    print('Please choose a sample size larger than 100: ')\n",
    "    sampleSize = int(input('Sample Size # = '))    \n",
    "\n",
    "    print('Would you like to use test data?')\n",
    "    randomize = input('Y or N? \\n')\n",
    "    #output will return 1 or 2 with the tuple of map distances. 1/2 will indicate the type of cross \n",
    "    data = []\n",
    "    if randomize == 'Y' or randomize == 'y':\n",
    "        A = random.randint(0,13)\n",
    "        B = random.randint(14,26)\n",
    "        C = random.randint(27,40)\n",
    "        D = random.randint(40,50)\n",
    "        #interference = random.uniform(0,1)\n",
    "        data = [A, B, C, D]\n",
    "    else:\n",
    "        A = int(input('A: '))\n",
    "        B = int(input('B: '))\n",
    "        C = int(input('C: '))\n",
    "        D = int(input('D: '))\n",
    "        #interference = int(input('interference: '))\n",
    "        data = [A, B, C, D]\n",
    "    return (crossType, data, sampleSize)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def mapDistChr1(chromosome1):\n",
    "    '''[Matt] input chromosome 1 to calculate the Map Units between each gene. Supports 2 or 3 genes'''\n",
    "    if len(chromosome1) == 3:\n",
    "        #3-1 cross\n",
    "        AB = abs(chromosome1[0] - chromosome1[1])\n",
    "        BC = abs(chromosome1[1] - chromosome1[2])\n",
    "        return [AB, BC]\n",
    "    elif len(chromosome1) == 2:\n",
    "        #2-2 cross\n",
    "        AB = abs(chromosome1[0] - chromosome1[1])\n",
    "        return AB\n",
    "\n",
    "def progenyNum(recombFrequency, N):\n",
    "    kids = []\n",
    "    for n in recombFrequency:\n",
    "        kids.append(int(n*N))\n",
    "    nonRecombinant = (.5 * N) - sum(kids)\n",
    "    kids.append(nonRecombinant)\n",
    "    return kids\n",
    "\n",
    "#in execute, if initialize[0] == 1, call this function\n",
    "def cross31(data, sampleSize): #take in second item (the list) from the initialize function\n",
    "    A = data[0]\n",
    "    B = data[1]\n",
    "    C = data[2]\n",
    "    D = data[3]\n",
    "\n",
    "    N = sampleSize\n",
    "\n",
    "    chromosome1 = [A, B, C]\n",
    "    chromosome2 = D\n",
    "\n",
    "    mapDist1 = mapDistChr1(chromosome1)\n",
    "    mapDist2 = 0\n",
    "\n",
    "    AB = mapDist1[0]\n",
    "    BC = mapDist1[1]\n",
    "    recomb1 = (AB / 100) * (1 - (BC / 100))\n",
    "    recomb2 = (BC / 100) * (1 - (AB / 100))\n",
    "    recomb3 = (AB / 100) * (BC / 100)\n",
    "\n",
    "    recombinants = [ recomb1, recomb2, recomb3 ]\n",
    "\n",
    "    prog = progenyNum(recombinants, N)\n",
    "\n",
    "    case1 = tabulate(\n",
    "        [\n",
    "            ['ABC', 'nonrecombinant', prog[3]], \n",
    "            ['abc','nonrecombinant', prog[3]],\n",
    "            ['aBC','single crossover AB',prog[0]],\n",
    "            ['Abc','single crossover AB',prog[0]],\n",
    "            ['ABc','single crossover BC',prog[1]],\n",
    "            ['abC','single crossover BC',prog[1]],\n",
    "            ['AbC','double crossover',prog[2]],\n",
    "            ['aBc','double crossover',prog[2]]\n",
    "            ], \n",
    "            headers=['Genotype', 'Type','Number of Progeny'])\n",
    "\n",
    "    return case1\n",
    "\n",
    "def cross22(data, sampleSize):\n",
    "    A = data[0]\n",
    "    B = data[1]\n",
    "    C = data[2]\n",
    "    D = data[3]\n",
    "\n",
    "    N = sampleSize\n",
    "\n",
    "    chromosome1 = [A, B]\n",
    "    chromosome2 = [C, D]\n",
    "\n",
    "    mapDist1 = mapDistChr1(chromosome1)\n",
    "    mapDist2 = abs(chromosome2[0] - chromosome2[1])\n",
    "\n",
    "    AB = mapDist1\n",
    "    CD = mapDist2\n",
    "\n",
    "    CDcross = (CD / 100) * (1 - (AB / 100))\n",
    "    ABcross = (AB / 100) * (1 - (CD / 100))\n",
    "    ABCDcross = (AB / 100) * (CD / 100)\n",
    "\n",
    "    recombinants = [CDcross, ABcross, ABCDcross]\n",
    "    \n",
    "    prog = []\n",
    "    for n in recombinants:\n",
    "        prog.append(int(.5*n*N))\n",
    "    nonRecombinant = (.25 * N) - sum(prog)\n",
    "    prog.append(nonRecombinant)\n",
    "\n",
    "    case2 = tabulate(\n",
    "        [\n",
    "            ['AB-CD', 'nonrecombinant', prog[3]], \n",
    "            ['AB-cd','nonrecombinant', prog[3]],\n",
    "            ['ab-cd','nonrecombinant',prog[3]],\n",
    "            ['ab-cD','nonrecombinant',prog[3]],\n",
    "            ['AB-Cd','CD crossover', prog[0]],\n",
    "            ['AB-cD','CD crossover', prog[0]],\n",
    "            ['ab-Cd','CD crossover', prog[0]],\n",
    "            ['ab-cD','CD crossover', prog[0]],\n",
    "            ['Ab-CD','AB crossover', prog[1]],\n",
    "            ['Ab-cd','AB crossover', prog[1]],\n",
    "            ['aB-CD','AB crossover', prog[1]],\n",
    "            ['aB-cd','AB crossover', prog[1]],\n",
    "            ['Ab-Cd','AB/CD crossover', prog[2]],\n",
    "            ['Ab-cD','AB/CD crossover', prog[2]],\n",
    "            ['aB-Cd','AB/CD crossover', prog[2]],\n",
    "            ['aB-cD','AB/CD crossover', prog[2]]\n",
    "            ], \n",
    "            headers=['Genotype', 'Type','Number of Progeny'])\n",
    "    return case2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to the genetic mapping simulator.\n",
      "What type of genetic cross would you like to observe?\n",
      "Option 1: 3-1 cross\n",
      "Press 1 for Option 1\n",
      "or\n",
      "Option 2: 2-2 cross\n",
      "Press 2 for Option 2\n"
     ]
    }
   ],
   "source": [
    "def execute():\n",
    "\n",
    "    init = initialize()\n",
    "\n",
    "    sampleSize = init[2]\n",
    "    data = init[1]\n",
    "    crossType = init[0]\n",
    "\n",
    "    if crossType == 1:\n",
    "        print(cross31(data, sampleSize))\n",
    "\n",
    "    elif crossType == 2:\n",
    "        print(cross22(data, sampleSize))\n",
    "\n",
    "execute()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "We create a dictionary that allows infinite keywords in case we want to expand the code. Right now we assume that if three are linked it is ABC and if two and two are linked its AB and CD. These functions then sort them accordingly.\n",
    "\n",
    "The first step is to create the dictionary "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
